### PR TITLE
client: try to invalidate the dcache when any shared caps lost

### PR DESCRIPTION
**Pull request by lxbsz**
We don't know what has happened when the MDS is revoking the shared
caps, maybe deleting the files, etc. We need to invalidate the dcaches
for the dentries linked to that inode to force to lookup it from the backend
in future.

Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by lxbsz** [comment]
jenkins retest this please

**Comment by lxbsz** [comment]
jenkins test make check

File: src/client/Client.cc [hunk 1]
```diff
@@ -5416,9 +5416,13 @@ void Client::_try_to_trim_inode(Inode *in, bool sched_inval)
...
         //        so in->dentries doesn't always reflect the state of kernel's dcache.
-        _schedule_invalidate_dentry_callback(dn, true);
+	if (!in->nlink)
+          _schedule_invalidate_dentry_callback(dn, true);
+	else
+          _schedule_invalidate_dentry_callback(dn, false);
       }
-      unlink(dn, true, true);
+      if (!in->nlink)
+        unlink(dn, true, true);
```
Comment (batrick) [inline src/client/Client.cc:5425]: These changes do not look related to `client: try to invalidate the dcache when any shared caps lost`. Can you explain? Perhaps in another commit?

Also, do you have a reproducer for the problem or is this something you believe exists but can't reproduce?

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/client/Client.cc (hunk 1 above)
Comment (lxbsz) [reply src/client/Client.cc:5425]: > These changes do not look related to `client: try to invalidate the dcache when any shared caps lost`. Can you explain? Perhaps in another commit?

I intend to invalidate the dentry cache only, in such as kernel fuse or third part app, like the Ganesha's NFS, when any shared caps gets lost. To make sure that the kernel fuse or Ganesha's NFS to ignore the relevant dentry caches. 

There has two cases will trigger this in `handle_cap_grant()`, the first one is when `nlink == 0` and another is when any shared caps gets lost.

> 
> Also, do you have a reproducer for the problem or is this something you believe exists but can't reproduce?

I didn't have the Ganesha NFS setups, so tried to use the ceph-fuse by hard coding to enable the `entry_timeout` options for the lookup today. The test case is :

```
mounter-cephfs-A:   'touch /mnt/cephfs-A/file0; ll /mnt/cephfs-A/'
mounter-cpehfs-B:   'ln /mnt/cephfs-B/file0 /mnt/cephfs-B/file1; rm /mnt/cephfs-B/file0 -f'
mounter-cephfs-A:   'll /cephfs/file0'
```
Without this PR, in the last step in `mounter-cephfs-A` we can still see the `file0`.

More detail please see: https://tracker.ceph.com/issues/53730#note-5

IMO, this should be enough to help `Sheng Xie` to push the above tracker to add the `entry_timeout` and `attr_timeout` support, which really could improve the perf much.

File: src/client/Client.cc [hunk 2]
```diff
@@ -5571,7 +5576,7 @@ void Client::handle_cap_grant(MetaSession *session, Inode *in, Cap *cap, const M
 
   // may drop inode's last ref
   if (deleted_inode)
-    _try_to_trim_inode(in, true);
+    _try_to_trim_inode(in, true, in->nlink == 0);
```
Comment (gregsfortytwo) [inline src/client/Client.cc:5579]: I'm not sure this condition works to preserve prior behavior — if there are hard links, we might have dropped *our* last reference to the inode without it ending up at nlink==0.

Comment (lxbsz) [reply src/client/Client.cc:5579]: Before only when `in->nlink == 0` will the `deleted_inode` be set. The code:

 ```
5455     if (in->nlink == 0)
5456       deleted_inode = true; 
```

And currently I added another case when `CEPH_STAT_CAP_INODE_ALL` caps are being revoked will the `deleted_inode` to be set too.

Yeah, I will push another version to make the code more readable.

**Review by rishabh-d-dave — COMMENTED** [review]
Looks okay to me.

**Review by rishabh-d-dave — APPROVED** [review]
Looks okay to me.

(Forgot to approve last time)

**Comment by rishabh-d-dave** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
jenkins test docs

**Comment by rishabh-d-dave** [comment]
@lxbsz Should this PR be put through QA?

**Comment by lxbsz** [comment]
> @lxbsz Should this PR be put through QA?

Yeah, please. Thanks.

**Comment by lxbsz** [comment]
Just rebased it and nothing changed.
